lib/new-statemachine.fnl

```diff
@@ -0,0 +1,212 @@
+(local atom (require :lib.atom))
+(local {: butlast
+        : call-when
+        : concat
+        : conj
+        : last
+        : merge
+        : slice} (require :lib.functional))
+
+(local log (hs.logger.new "\tstatemachine.fnl\t" "debug"))
+
+;; Finite state machine
+;; Schema
+;; {
+;;  ; An atom keyword
+;;  :current-state :state1
+;;  ; States table: A map of state names to a map of actions to functions
+;;  ; These functions must return a map containing the new state keyword, the
+;;  ; effect, and a new context
+;;  :states {:state1 {}
+;;           :state2 {}
+;;           :state3 {:leave state3-leave
+;;                    :exit state3-exit}}
+;;  ; An atom table
+;;  :context {}}
+
+
+(fn set-current-state
+  [fsm state]
+  (atom.swap! fsm.current-state (fn [_ state] state) state))
+
+(fn set-context
+  [fsm context]
+  (atom.swap! fsm.context (fn [_ context] context) context))
+
+(fn signal
```

Comment:
As mentioned in your own comments, might be helpful to break this up so that there's one function to determine if a transition is possible, and another to perform the transition, update the state atom, and publish the transition to subscribers

Reply:
What do you propose we do when a signal arrives that has no function registered? That's a bug in the specific state machine, no?

Reply:
My inclination is to just emit a warning using the Hammerspoon logging api https://www.hammerspoon.org/docs/hs.logger.html, possibly info or debug level. 

My reasoning is that if we throw an error, then we're contradicting the purpose of a statemachine because now consumers would need to imperatively check the state before publishing, which is not the code we generally want to spend our time writing.

A warning at the debug or info level will be hidden for regular users, but we can turn it on in our personal configs or temporarily while working on development.

I see a few implementation paths:

1. Create a general statemachine logger instance, expose it with exports so we can customize the log level in our configs
2. Create a logger instance for every state machine instance, then could update the level on a case-by-case basis using `(tset fsm.logger :level 5)`
3. Since adopting the init table for the `create-machine` api, it could take an extra boolean attribute such as `:log true` to log invalid transitions (options like this are why I recommended the table arg 😅 )
4. Could create an advisable function that's called with the fsm instance, current state, action, extra when it cannot transition that we could either opt-into logging it with advice or opt-out

Reply:
For now I did basically #3 but instead of taking a boolean it takes a string used to prepend log messages

Reply:
That works, would it be worth having a `:name` or `:id` field do you think?

Reply:
something like
`{:log {:name x}}`. Why is that?

Reply:
Was thinking like:

```fennel
{:name :modal-fsm
 :state {:current-state :closed
             :context {...}
 :states {...}
 :log      true}
```

Plus surfacing the name in the transition object sent to subscribers... but that also gets into what-if engineering territory so may not be valuable.

Reply:
```lisp
each [_ sub (pairs (atom.deref fsm.subscribers))]
  (sub {:prev-state state :next-state new-state : action : effect : extra})
```

For future proofing it's good we provide a dictionary here. We can add extra keys.

Note that in MY NAMING this isn't a transition object. A transition object is what an transition function returns, which is just a map with a new state and an effect. It would probably be helpful to either use standard language (and I am making up words as I go) or at least get on the same page. Any suggestion for what we'd call 1. The object returned from a transition function (what I call transition object) and 2. What we pass to the subscribers?

Any, I see what you mean now. You're suggesting allowing an ID/name field to the constructor, IF it's provided, use it in the logger AND maybe provide it to transition functions and subscribers. Correct?

Reply:
> ```lisp
> each [_ sub (pairs (atom.deref fsm.subscribers))]
>   (sub {:prev-state state :next-state new-state : action : effect : extra})
> ```
> For future proofing it's good we provide a dictionary here. We can add extra keys.
>

Oh! I did not catch that it wasn't a table. I agree, making it a table would be wise here for supporting changes over time and destructuring.

> Note that in MY NAMING this isn't a transition object. A transition object is what a transition function returns, which is just a map with a new state and an effect. It would probably be helpful to either use standard language (and I am making up words as I go) or at least get on the same page. Any suggestion for what we'd call 1. The object returned from a transition function (what I call transition object) and 2. What we pass to the subscribers?
>

That's a good point. Thoughts on this?

```
(signal :action extra?) -> (transition-fn ...) -> transition -> (sub transition-record)`
```

> Any, I see what you mean now. You're suggesting allowing an ID/name field to the constructor, IF it's provided, use it in the logger AND maybe provide it to transition functions and subscribers. Correct?
>

Correct

Reply:
> That's a good point. Thoughts on this?
> 
> ```
> (signal :action extra?) -> (transition-fn ...) -> transition -> (sub transition-record)`
> ```

So signal takes an extra, passes it to a transition function, which returns a transition object, which is then packed into a transition record for its subscribers? That makes sense to me. We can use that nomenclature in the docs.

Reply:
Exactly!